File: julia/tensor.jl
# SPDX-License-Identifier: CC-BY-NC-4.0
# Copyright (c) 2025-2026 fumi-engineer

# tensor.jl — Core tensor type and element-wise / BLAS operations
#
# All data is stored as Array{Float32} regardless of the logical DType tag.
# The DType enum exists for future mixed-precision support; currently F32 only.
#
# Key design decisions:
#   - Julia is column-major: a (M, K) matrix has M rows and K columns, with
#     each column of M elements stored contiguously (first index varies fastest).
#     This is the natural layout for BLAS (Fortran-order), so mul! on 2D
#     matrices dispatches directly to sgemm with no copy.
#   - All mul! calls go through LinearAlgebra, which dispatches via LBT
#     (libblastrampoline). On macOS with AppleAccelerate.jl loaded, this
#     routes to Apple's Accelerate framework (AMX hardware), giving 7-14x
#     speedup over NEON-only OpenBLAS.
#   - In-place (!-suffix) functions mutate their first argument to avoid
#     allocation. The forward pass pre-allocates all buffers so the hot
#     path triggers 0 GC pauses.

@enum DType F32 F16 BF16 I32 I64

mutable struct Tensor
    data::Array{Float32}
    dtype::DType
    grad::Union{Array{Float32}, Nothing}
end

# Convenience constructor: default grad to nothing (backward-compatible)
Tensor(data::Array{Float32}, dtype::DType) = Tensor(data, dtype, nothing)

# --- Constructors -----------------------------------------------------------

zeros_tensor(dims::Int...; dtype::DType=F32) = Tensor(zeros(Float32, dims...), dtype)
ones_tensor(dims::Int...; dtype::DType=F32) = Tensor(ones(Float32, dims...), dtype)
randn_tensor(dims::Int...; dtype::DType=F32) = Tensor(randn(Float32, dims...), dtype)
randn_std(dims::Int...; std::Float32=1f0, dtype::DType=F32) = Tensor(randn(Float32, dims...) .* std, dtype)
from_array(arr::AbstractArray; dtype::DType=F32) = Tensor(Float32.(arr), dtype)

Base.size(t::Tensor) = size(t.data)
numel(t::Tensor) = length(t.data)
clone(t::Tensor) = Tensor(copy(t.data), t.dtype, t.grad === nothing ? nothing : copy(t.grad))

# --- Element-wise binary ops (allocating) ------------------------------------

Base.:+(a::Tensor, b::Tensor) = _tensor_binop_add(a.data, b.data, a.dtype)
Base.:-(a::Tensor, b::Tensor) = _tensor_binop_sub(a.data, b.data, a.dtype)
Base.:*(a::Tensor, b::Tensor) = _tensor_binop_mul(a.data, b.data, a.dtype)

function _tensor_binop_add(a::Array{Float32,N}, b::Array{Float32,M}, dtype::DType) where {N,M}
    Tensor(a .+ b, dtype)
end
function _tensor_binop_sub(a::Array{Float32,N}, b::Array{Float32,M}, dtype::DType) where {N,M}
    Tensor(a .- b, dtype)
end
function _tensor_binop_mul(a::Array{Float32,N}, b::Array{Float32,M}, dtype::DType) where {N,M}
    Tensor(a .* b, dtype)
end

scale(t::Tensor, s) = _tensor_scale(t.data, Float32(s), t.dtype)
function _tensor_scale(d::Array{Float32,N}, s::Float32, dtype::DType) where {N}
    Tensor(d .* s, dtype)
end

# SiLU (Swish): f(x) = x * sigmoid(x) = x / (1 + exp(-x))
silu(t::Tensor) = _tensor_silu(t.data, t.dtype)
function _tensor_silu(d::Array{Float32,N}, dtype::DType) where {N}
    Tensor(@.(d / (1f0 + exp(-d))), dtype)
end

# SiLU: f(x) = x / (1 + exp(-x))
# In-place variant: mutates data array directly, avoids allocation.
# @inbounds is safe here: eachindex always yields valid indices.
# @simd enables SIMD vectorization for the simple element-wise loop.
function silu_in_place!(t::Tensor)
    _silu_in_place_arr!(t.data)
    return t
end

function _silu_in_place_arr!(d::Array{Float32,N}) where {N}
    @inbounds @simd for i in eachindex(d)
        x = d[i]
        d[i] = x / (1f0 + exp(-x))
    end
end

# --- In-place element-wise ops (zero-alloc) ----------------------------------

function add_in_place!(a::Tensor, b::Tensor)
    _add_in_place_arr!(a.data, b.data)
    return a
end

function _add_in_place_arr!(ad::Array{Float32,N}, bd::Array{Float32,M}) where {N,M}
    ad .+= bd
end

function scale_in_place!(t::Tensor, s)
    _scale_in_place_arr!(t.data, Float32(s))
    return t
end

function _scale_in_place_arr!(d::Array{Float32,N}, s::Float32) where {N}
    d .*= s
end

# In-place element-wise multiply: a .*= b
function mul_in_place!(a::Tensor, b::Tensor)
    _mul_in_place_arr!(a.data, b.data)
    return a
end

function _mul_in_place_arr!(ad::Array{Float32,N}, bd::Array{Float32,M}) where {N,M}
    ad .*= bd
end

tensor_sum(t::Tensor)::Float32 = _tensor_sum(t.data)
tensor_mean(t::Tensor)::Float32 = _tensor_sum(t.data) / Float32(length(t.data))

_tensor_sum(d::Array{Float32,N}) where {N} = sum(d)

# --- Matrix multiplication ---------------------------------------------------
# 2D matmul: C = A * B  where A is (M, K), B is (K, N), C is (M, N)
# Julia column-major: columns are contiguous, so (M, K) has stride-1 along M.
# mul! dispatches to BLAS sgemm via LBT -> AppleAccelerate (AMX) on macOS.

function tensor_matmul(a::Tensor, b::Tensor, ::Val{2})::Tensor
    _tensor_matmul_2d(a.data, b.data, a.dtype)
end

function _tensor_matmul_2d(ad::Array{Float32,N1}, bd::Array{Float32,N2}, dtype::DType) where {N1,N2}
    M, K = size(ad)
    K2, N = size(bd)
    K == K2 || error("matmul dimension mismatch: $K vs $K2")
    ad2 = reshape(ad, M, K)
    bd2 = reshape(bd, K, N)
    out = Matrix{Float32}(undef, M, N)
    mul!(out, ad2, bd2)  # BLAS sgemm via LBT
    return Tensor(out, dtype)
end

# 3D batched matmul: C[b,:,:] = A[b,:,:] * B[b,:,:]
# Julia stores 3D arrays in column-major order: the first index varies fastest.
# For a (B, M, K) array, @view arr[batch,:,:] yields a non-contiguous slice
# with stride[1]=B (not 1). BLAS sgemm requires stride[1]=1 (contiguous columns),
# so we must copy each batch slice into contiguous 2D buffers before calling mul!.
# Without this copy, Julia would fall back to generic matmul (~10x slower).
function tensor_matmul(a::Tensor, b::Tensor, ::Val{3})::Tensor
    _tensor_matmul_3d(a.data, b.data, a.dtype)
end

function _tensor_matmul_3d(ad::Array{Float32,N1}, bd::Array{Float32,N2}, dtype::DType) where {N1,N2}
    B, M, K = size(ad)
    B2, K2, N = size(bd)
    B == B2 || error("batch size mismatch: $B vs $B2")
    K == K2 || error("matmul dimension mismatch: $K vs $K2")
    out = Array{Float32}(undef, B, M, N)
    # Pre-allocate contiguous 2D buffers for BLAS-compatible slices.
    # @view ad[batch,:,:] has stride[1]=B (non-contiguous) -- BLAS sgemm requires
    # stride[1]=1. Copy each batch slice into contiguous buffers for real BLAS dispatch.
    a_buf = Matrix{Float32}(undef, M, K)
    b_buf = Matrix{Float32}(undef, K, N)
    o_buf = Matrix{Float32}(undef, M, N)
    for batch in 1:B
        # Manual copy: column-major iteration (inner=rows, outer=cols) for cache locality
        @inbounds for j in 1:K, i in 1:M
            a_buf[i, j] = ad[batch, i, j]
        end
        @inbounds for j in 1:N, i in 1:K
            b_buf[i, j] = bd[batch, i, j]
        end
        mul!(o_buf, a_buf, b_buf)  # BLAS sgemm on contiguous data
        @inbounds for j in 1:N, i in 1:M
            out[batch, i, j] = o_buf[i, j]
        end
    end
    return Tensor(out, dtype)
end

tensor_matmul(::Tensor, ::Tensor, ::Val{N}) where {N} = error("matmul not implemented for $(N)D tensors")
tensor_matmul(a::Tensor, b::Tensor) = tensor_matmul(a, b, Val(ndims(a.data)))

# C = A * B^T  where A is (M, K), B is (N, K), C is (M, N)
# transpose(bd2) creates a lazy wrapper; mul! fuses the transpose into sgemm (TRANSA/TRANSB flags).
function matmul_transposed_b(a::Tensor, b::Tensor)::Tensor
    _matmul_transposed_b_arr(a.data, b.data, a.dtype)
end

function _matmul_transposed_b_arr(ad::Array{Float32,N1}, bd::Array{Float32,N2}, dtype::DType) where {N1,N2}
    M, K = size(ad)
    N, K2 = size(bd)
    K == K2 || error("matmulT dimension mismatch: $K vs $K2")
    ad2 = reshape(ad, M, K)
    bd2 = reshape(bd, N, K)
    out = Matrix{Float32}(undef, M, N)
    mul!(out, ad2, transpose(bd2))  # sgemm with TRANSB='T'
    Tensor(out, dtype)
end

# In-place matmul_transposed_b: writes result into pre-allocated output (Tensor args)
function matmul_transposed_b!(out::Matrix{Float32}, a::Tensor, b::Tensor)
    mul!(out, a.data, transpose(b.data))
    return out
end

# In-place matmul_transposed_b on raw Arrays
function matmul_transposed_b!(out::Matrix{Float32}, a::Matrix{Float32}, b::Matrix{Float32})
    mul!(out, a, transpose(b))
    return out
end

# In-place 2D matmul: out = a * b, writing into pre-allocated out
function tensor_matmul!(out::Matrix{Float32}, a::Matrix{Float32}, b::Matrix{Float32})
    mul!(out, a, b)
    return out
end

# --- Softmax -----------------------------------------------------------------
# Softmax: p_i = exp(x_i - max(x)) / sum(exp(x_j - max(x)))
# The max-subtraction trick prevents overflow in exp().
# Three-pass algorithm: (1) find max, (2) exp + sum, (3) normalize.

# In-place softmax for 2D arrays: operates on reshaped views.
# Each row is treated as an independent probability vector.
function softmax!(out::AbstractMatrix{Float32}, src::AbstractMatrix{Float32})
    num_vectors, last_dim = size(src)
    for v in 1:num_vectors
        mx = -Inf32
        @inbounds for i in 1:last_dim
            val = src[v, i]
            mx = ifelse(val > mx, val, mx)
        end
        s = 0f0
        @inbounds for i in 1:last_dim
            e = exp(src[v, i] - mx)
            out[v, i] = e
            s += e
        end
        inv_s = 1f0 / s
        @inbounds @simd for i in 1:last_dim
            out[v, i] *= inv_s
        end
    end
    out
end

# In-place element-wise add on raw arrays: a .+= b
function add!(a::Array{Float32}, b::Array{Float32})
    a .+= b
    return a
end

# In-place element-wise subtract on raw arrays: a .-= b
function sub!(a::Array{Float32}, b::Array{Float32})
    a .-= b
    return a
end

# Transpose: swaps the last two dimensions via permutedims.
# permutedims allocates a new array (not a view).
function transpose_tensor(t::Tensor)::Tensor
    _transpose_arr(t.data, t.dtype)
end

function _transpose_arr(d::Array{Float32,N}, dtype::DType) where {N}
    N >= 2 || error("transpose requires at least 2D tensor")
    perm = collect(1:N)
    perm[end], perm[end-1] = perm[end-1], perm[end]
    Tensor(permutedims(d, perm), dtype)
end

# reshape is a zero-copy view in Julia (shares underlying memory).
reshape_tensor(t::Tensor, dims::Int...) = Tensor(reshape(t.data, dims...), t.dtype)

# Softmax: p_i = exp(x_i - max(x)) / sum(exp(x_j - max(x)))
# Allocating version: creates new output array.
# Flattens to (num_vectors, last_dim) for uniform iteration.
function softmax(t::Tensor)::Tensor
    _softmax_arr(t.data, t.dtype)
end

function _softmax_arr(src::Array{Float32,N}, dtype::DType)::Tensor where {N}
    nd = N
    nd >= 1 || error("softmax requires at least 1 dimension")
    out = similar(src)
    last_dim = size(src, nd)
    num_vectors = div(length(src), last_dim)
    flat_src = reshape(src, num_vectors, last_dim)
    flat_dst = reshape(out, num_vectors, last_dim)
    for v in 1:num_vectors
        mx = -Inf32
        @inbounds for i in 1:last_dim
            val = flat_src[v, i]
            mx = ifelse(val > mx, val, mx)
        end
        s = 0f0
        @inbounds for i in 1:last_dim
            e = exp(flat_src[v, i] - mx)
            flat_dst[v, i] = e
            s += e
        end
        inv_s = 1f0 / s
        @inbounds @simd for i in 1:last_dim
            flat_dst[v, i] *= inv_s
        end
    end
    Tensor(out, dtype)
end

# Softmax: p_i = exp(x_i - max(x)) / sum(exp(x_j - max(x)))
# In-place into a pre-allocated output buffer (same shape as input).
# Concrete dispatch on Array{Float32,N} avoids abstract-type penalty.
function softmax_into!(out::Array{Float32,N}, src::Array{Float32,M}, dtype::DType)::Tensor where {N,M}
    last_dim = size(src, M)
    num_vectors = div(length(src), last_dim)
    flat_src = reshape(src, num_vectors, last_dim)
    flat_dst = reshape(out, num_vectors, last_dim)
    for v in 1:num_vectors
        mx = -Inf32
        @inbounds for i in 1:last_dim
            val = flat_src[v, i]
            mx = ifelse(val > mx, val, mx)
        end
        s = 0f0
        @inbounds for i in 1:last_dim
            e = exp(flat_src[v, i] - mx)
            flat_dst[v, i] = e
            s += e
        end
        inv_s = 1f0 / s
        @inbounds @simd for i in 1:last_dim
            flat_dst[v, i] *= inv_s
        end
    end
    Tensor(out, dtype)
end

# Softmax: p_i = exp(x_i - max(x)) / sum(exp(x_j - max(x)))
# In-place on a 1D vector. Used for attention score rows.
function softmax_in_place!(xs::AbstractVector{Float32})
    mx = -Inf32
    @inbounds for i in eachindex(xs)
        val = xs[i]
        mx = ifelse(val > mx, val, mx)
    end
    s = 0f0
    @inbounds for i in eachindex(xs)
        e = exp(xs[i] - mx)
        xs[i] = e
        s += e
    end
    inv_s = 1f0 / s
    @inbounds @simd for i in eachindex(xs)
        xs[i] *= inv_s
    end
    return xs
end

argmax_f32(xs::AbstractVector{Float32})::Tuple{Int,Float32} = reverse(findmax(xs))

# L1 normalization: x_i = x_i / sum(x)
function normalize_in_place!(xs::AbstractVector{Float32})
    s = 0f0
    @inbounds for i in eachindex(xs)
        s += xs[i]
    end
    s == 0f0 && return xs
    inv_s = 1f0 / s
    @inbounds @simd for i in eachindex(xs)
        xs[i] *= inv_s
    end
    return xs
end

File: python/tensor.py
# SPDX-License-Identifier: CC-BY-NC-SA-4.0
# Copyright (c) 2025-2026 fumi-engineer

"""Tensor operations for MoE Transformer.

Thin wrapper around numpy ndarrays providing a typed Tensor API.
On macOS >= 14, numpy's np.matmul dispatches to Apple Accelerate (BLAS/AMX),
so all matmul calls in this project benefit from hardware-accelerated linear
algebra without any explicit FFI.
"""

from __future__ import annotations

from enum import Enum
from typing import Sequence

import numpy as np


class DType(Enum):
    """Supported data types for tensors.

    Maps our enum variants to numpy dtypes.  BF16 is emulated as float32
    because numpy has no native bfloat16 support.
    """

    F32 = "float32"
    F16 = "float16"
    BF16 = "bfloat16"
    I32 = "int32"
    I64 = "int64"

    def to_numpy(self) -> np.dtype:
        """Convert to numpy dtype."""
        if self == DType.BF16:
            # numpy doesn't have bfloat16, use float32 as a stand-in
            return np.float32
        return np.dtype(self.value)


class Tensor:
    """Tensor class wrapping numpy arrays.

    Design notes:
    - __init__ avoids a copy when the source array already has the target
      dtype (``data.dtype == target``).  This matters in hot paths where
      Tensor.from_numpy is called with pre-cast arrays.
    - All arithmetic methods return new Tensor instances (immutable style).
    """

    def __init__(self, data: np.ndarray, dtype: DType = DType.F32):
        """Create tensor from numpy array.

        Skips the dtype cast (and therefore the copy) when ``data.dtype``
        already matches the target numpy dtype.  This is intentional: most
        internal code constructs arrays in the correct dtype up-front, so
        the branch avoids a redundant allocation on every layer output.
        """
        target = dtype.to_numpy()
        # Skip copy when dtype already matches — avoids allocation overhead
        self._data = data if data.dtype == target else data.astype(target)
        self._dtype = dtype
        # Gradient storage for backward pass — set by layer backward methods
        self._grad: np.ndarray | None = None

    @property
    def shape(self) -> tuple[int, ...]:
        """Return tensor shape."""
        return self._data.shape

    @property
    def dtype(self) -> DType:
        """Return tensor dtype."""
        return self._dtype

    @property
    def numel(self) -> int:
        """Return number of elements."""
        return self._data.size

    @property
    def data(self) -> np.ndarray:
        """Return underlying numpy array."""
        return self._data

    @classmethod
    def zeros(cls, shape: Sequence[int], dtype: DType = DType.F32) -> Tensor:
        """Create tensor of zeros."""
        return cls(np.zeros(shape, dtype=dtype.to_numpy()), dtype)

    @classmethod
    def ones(cls, shape: Sequence[int], dtype: DType = DType.F32) -> Tensor:
        """Create tensor of ones."""
        return cls(np.ones(shape, dtype=dtype.to_numpy()), dtype)

    @classmethod
    def randn(cls, shape: Sequence[int], dtype: DType = DType.F32) -> Tensor:
        """Create tensor with random normal values (mean=0, std=1)."""
        return cls(np.random.randn(*shape).astype(dtype.to_numpy()), dtype)

    @classmethod
    def randn_std(
        cls, shape: Sequence[int], dtype: DType = DType.F32, std: float = 1.0
    ) -> Tensor:
        """Create tensor with random normal values and specific std."""
        return cls((np.random.randn(*shape) * std).astype(dtype.to_numpy()), dtype)

    @classmethod
    def from_numpy(cls, arr: np.ndarray, dtype: DType = DType.F32) -> Tensor:
        """Create tensor from numpy array."""
        return cls(arr, dtype)

    def clone(self) -> Tensor:
        """Create a deep copy of the tensor."""
        return Tensor(self._data.copy(), self._dtype)

    def reshape(self, shape: Sequence[int]) -> Tensor:
        """Reshape tensor (returns a view when possible, copy otherwise)."""
        return Tensor(self._data.reshape(shape), self._dtype)

    def transpose(self, axes: Sequence[int] | None = None) -> Tensor:
        """Transpose tensor."""
        if axes is None:
            # Default: reverse all axes
            axes = tuple(range(len(self.shape) - 1, -1, -1))
        return Tensor(np.transpose(self._data, axes), self._dtype)

    def add(self, other: Tensor) -> Tensor:
        """Element-wise addition."""
        return Tensor(self._data + other._data, self._dtype)

    def sub(self, other: Tensor) -> Tensor:
        """Element-wise subtraction."""
        return Tensor(self._data - other._data, self._dtype)

    def mul(self, other: Tensor) -> Tensor:
        """Element-wise multiplication."""
        return Tensor(self._data * other._data, self._dtype)

    def div(self, other: Tensor) -> Tensor:
        """Element-wise division."""
        return Tensor(self._data / other._data, self._dtype)

    def scale(self, scalar: float) -> Tensor:
        """Scale tensor by scalar."""
        return Tensor(self._data * scalar, self._dtype)

    def neg(self) -> Tensor:
        """Negate tensor."""
        return Tensor(-self._data, self._dtype)

    def exp(self) -> Tensor:
        """Element-wise exponential."""
        return Tensor(np.exp(self._data), self._dtype)

    def log(self) -> Tensor:
        """Element-wise natural logarithm."""
        return Tensor(np.log(self._data), self._dtype)

    def sqrt(self) -> Tensor:
        """Element-wise square root."""
        return Tensor(np.sqrt(self._data), self._dtype)

    def pow(self, exponent: float) -> Tensor:
        """Element-wise power."""
        return Tensor(np.power(self._data, exponent), self._dtype)

    def sum(self, axis: int | None = None, keepdims: bool = False) -> Tensor:
        """Sum along axis."""
        return Tensor(np.sum(self._data, axis=axis, keepdims=keepdims), self._dtype)

    def mean(self, axis: int | None = None, keepdims: bool = False) -> Tensor:
        """Mean along axis."""
        return Tensor(np.mean(self._data, axis=axis, keepdims=keepdims), self._dtype)

    def max(self, axis: int | None = None, keepdims: bool = False) -> Tensor:
        """Max along axis."""
        return Tensor(np.max(self._data, axis=axis, keepdims=keepdims), self._dtype)

    def argmax(self, axis: int = -1) -> Tensor:
        """Argmax along axis."""
        return Tensor(np.argmax(self._data, axis=axis), DType.I64)

    def silu(self) -> Tensor:
        """SiLU (Swish) activation.

        SiLU: y = x * sigmoid(x) = x / (1 + exp(-x))
        """
        sigmoid = 1.0 / (1.0 + np.exp(-self._data))
        return Tensor(self._data * sigmoid, self._dtype)

    def silu_inplace(self) -> Tensor:
        """In-place SiLU (Swish) activation. Returns self for chaining."""
        sigmoid = 1.0 / (1.0 + np.exp(-self._data))
        np.multiply(self._data, sigmoid, out=self._data)
        return self

    def mul_inplace(self, other: Tensor) -> Tensor:
        """In-place element-wise multiplication. Returns self for chaining."""
        np.multiply(self._data, other._data, out=self._data)
        return self

    def add_inplace(self, other: Tensor) -> Tensor:
        """In-place element-wise addition. Returns self for chaining."""
        np.add(self._data, other._data, out=self._data)
        return self

    def softmax(self, axis: int = -1) -> Tensor:
        """Softmax along axis.

        Softmax: p_i = exp(x_i - max(x)) / sum_j(exp(x_j - max(x)))

        Subtracting max(x) before exp() prevents overflow (numerical
        stability trick).  The result is mathematically identical because
        exp(x_i - c) / sum(exp(x_j - c)) = exp(x_i) / sum(exp(x_j)).
        """
        shifted = self._data - np.max(self._data, axis=axis, keepdims=True)
        exp_vals = np.exp(shifted)
        return Tensor(exp_vals / np.sum(exp_vals, axis=axis, keepdims=True), self._dtype)

    def matmul(self, other: Tensor) -> Tensor:
        """Matrix multiplication via NumPy/BLAS.

        On macOS with numpy linked against Accelerate, this dispatches to
        AMX-backed sgemm for float32 — no explicit FFI required.
        """
        return Tensor(np.matmul(self._data, other._data), self._dtype)

    def __add__(self, other: Tensor) -> Tensor:
        return self.add(other)

    def __sub__(self, other: Tensor) -> Tensor:
        return self.sub(other)

    def __mul__(self, other: Tensor | float) -> Tensor:
        if isinstance(other, Tensor):
            return self.mul(other)
        return self.scale(other)

    def __rmul__(self, other: float) -> Tensor:
        return self.scale(other)

    def __neg__(self) -> Tensor:
        return self.neg()

    def __matmul__(self, other: Tensor) -> Tensor:
        return self.matmul(other)

    def __repr__(self) -> str:
        return f"Tensor(shape={self.shape}, dtype={self._dtype.name})"

    def __getitem__(self, key):
        """Index into tensor."""
        result = self._data[key]
        if isinstance(result, np.ndarray):
            return Tensor(result, self._dtype)
        return result
